src/ripple/consensus/Validations.h

```diff
@@ -0,0 +1,504 @@
+                impl().onStale(std::move(it->second));
+                it = current_.erase(it);
+            }
+            else
+            {
+                auto cit = typename ValidationMap::const_iterator{it};
+                // contains a live record
+                f(cit->first, cit->second);
+                ++it;
+            }
+        }
+    }
+
+    /** Return set of validations for a ledger
+
+        Returns the set of validations for a ledger.
+        @param ledgerID The identifier of the ledger
+        @return A pointer to the set of validations, nullptr if none exist.
+    */
+    ValidationMap const*
+    byLedger(LedgerID const& ledgerID)
+    {
+        auto it = byLedger_.find(ledgerID);
+        if (it != byLedger_.end())
+        {
+            // Update set time since it is being used
+            byLedger_.touch(it);
+            return &(it->second);
+        }
+        return nullptr;
+    }
+
+    //! Return the validation timing parameters
+    ValidationParms const&
+    parms() const
+    {
+        return parms_;
+    }
+
+    //! Flush all current validations, calling `Derived::onStale` member.
+    void
+    flush()
+    {
+        for (auto& it : current_)
+        {
+            impl().onStale(std::move(it.second));
+        }
+        current_.clear();
+    }
+
+public:
+    /** Constructor
+
+        @param p ValidationParms to control staleness/expiration of validaitons
+        @param c Clock to use for expiring validations stored by ledger
+    */
+    Validations(
+        ValidationParms const& p,
+        beast::abstract_clock<std::chrono::steady_clock>& c)
+        : byLedger_(c), parms_(p)
+    {
+    }
+
+    /** Result of adding a new validation
+     */
+    enum class AddOutcome {
+        current,  //< This was a new validation and was added
+        repeat,   //< Already had this validation
+        stale,    //< Not current or was older than current from this node
+        sameSeq,  //< Had a validation with same sequence number
+    };
+
+    /** Add a new validation
+
+        Attempt to add a new validation.
+
+        @param t Current time
+        @param key The NodeKey to use for the validation
+        @param val The validation to store
+        @return The outcome of the attempt
+
+        @note The provided key may differ from the validation's
+              key() member since we might be storing by master key and the
+              validation might be signed by a temporary or rotating key.
+
+    */
+    AddOutcome
+    add(NetClock::time_point t, NodeKey const& key, Validation const& val)
+    {
+        if (!isCurrent(parms_, t, val.signTime(), val.seenTime()))
+            return AddOutcome::stale;
+
+        auto const& id = val.ledgerID();
+
+        if (!byLedger_[id].emplace(key, val).second)
+            return AddOutcome::repeat;
+
+        AddOutcome result = AddOutcome::current;
+
+        // Attempt to insert
+        auto ins = current_.emplace(key, val);
```

Comment:
If we need to replace the current validation because it is stale, we need to [modify `ins`](https://github.com/ripple/rippled/pull/2084/files/c4a607efb55e31bbcc66b5fa4ca91c68780b500f..03fbaadad201877c0c6ca13249d0d303517c1af5#diff-46f83c3cd15dd825b63068c0ab4dea00R369)